keps/sig-node/20190528-node-maintenance-controller.md

```diff
@@ -0,0 +1,190 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Node maintenance controller that performs node drain - server side using custom resources calls to the API in order to perform the drain (cordon + evictions)
+
+## Motivation
+
+Move kubectl drain into the server to enable execution of node drain using custom resources calls to the API in order to perform the drain.
+This would also reduce the proliferation of drain behaviours in other projects that re-implement the drain logic to accustom their needs.
+
+(https://github.com/kubernetes/kubernetes/issues/25625)
+
+### Goals
+
+- A new controller that implements the drain logic that is currently [implemented](https://github.com/kubernetes/kubernetes/blob/v1.14.0/pkg/kubectl/cmd/drain/drain.go#L307) in [kubectl](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#drain).
+- The controller watches and acts upon maintenance CRs to put/remove a node on/from maintenance.
+- Status report of the drain (embodied in the maintenance CR).
+- Custom taints and labels added to the node going under maintenance.
+- Configurable time-out(time buffer) for the drain process.
+
+### Non-Goals
+- "Dry run" mode to detect what is the best node / N nodes to drain
+
+## Proposal
+
+Node maintenance controller watches for new or deleted custom resources e.g. `NodeMaintenance` which indicate that a node in the cluster should either:
+ - `NodeMaintenance` CR created: move node into maintenance, cordon the node - set it as unscheduleble and evict the pods (which can be evicted) from that node.
+  - `NodeMaintenance` CR deleted: remove pod from maintenance and uncordon the node - set it as scheduleble.
+
+Current implementation and design can be found under Kubevirt:
+https://github.com/kubevirt/node-maintenance-operator 
+
+### User Stories
+
+#### Story 1
+As a third party developer/operator I want to evacuate my workloads appropriately when a node goes into maintenance mode, but I do not want to evacuate them when the node is cordoned.
+
+- Today we can not differentiate between cordon and drain using cluster information
```

Comment:
I myself don't understand the difference between cordon and drain. Why can we not tell the difference today? Please consider adding a description of the terms somewhere in this proposal.